exporters/otlp/src/main/java/io/opentelemetry/exporter/otlp/AttributeMarshaler.java

```diff
@@ -0,0 +1,293 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.exporter.otlp;
+
+import com.google.protobuf.ByteString;
+import com.google.protobuf.CodedOutputStream;
+import com.google.protobuf.WireFormat;
+import io.opentelemetry.api.common.AttributeKey;
+import io.opentelemetry.api.common.Attributes;
+import io.opentelemetry.proto.common.v1.AnyValue;
+import io.opentelemetry.proto.common.v1.ArrayValue;
+import io.opentelemetry.proto.common.v1.KeyValue;
+import java.io.IOException;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+
+abstract class AttributeMarshaler extends MarshalerWithSize {
+  private static final AttributeMarshaler[] EMPTY_REPEATED = new AttributeMarshaler[0];
+  private final ByteString key;
+  private final int valueSize;
+
+  static AttributeMarshaler[] createRepeated(Attributes attributes) {
+    if (attributes.isEmpty()) {
+      return EMPTY_REPEATED;
+    }
+
+    AttributeMarshaler[] attributeMarshalers = new AttributeMarshaler[attributes.size()];
+    // TODO: Revisit how to avoid the atomic integer creation.
+    AtomicInteger pos = new AtomicInteger();
+    attributes.forEach(
+        (attributeKey, o) ->
+            attributeMarshalers[pos.getAndIncrement()] =
+                AttributeMarshaler.create(attributeKey, o));
+    return attributeMarshalers;
+  }
+
+  @SuppressWarnings("unchecked")
+  static AttributeMarshaler create(AttributeKey<?> attributeKey, Object value) {
+    ByteString key = MarshalerUtil.toByteString(attributeKey.getKey());
+    if (value == null) {
+      return new KeyValueNullMarshaler(key);
+    }
+    switch (attributeKey.getType()) {
+      case STRING:
+        return new KeyValueStringMarshaler(key, MarshalerUtil.toByteString((String) value));
+      case LONG:
+        return new KeyValueLongMarshaler(key, (Long) value);
+      case BOOLEAN:
+        return new KeyValueBooleanMarshaler(key, (Boolean) value);
+      case DOUBLE:
+        return new KeyValueDoubleMarshaler(key, (Double) value);
+      case STRING_ARRAY:
+        return new KeyValueArrayStringMarshaler(key, (List<String>) value);
+      case LONG_ARRAY:
+        return new KeyValueArrayLongMarshaler(key, (List<Long>) value);
+      case BOOLEAN_ARRAY:
+        return new KeyValueArrayBooleanMarshaler(key, (List<Boolean>) value);
+      case DOUBLE_ARRAY:
+        return new KeyValueArrayDoubleMarshaler(key, (List<Double>) value);
+    }
+    throw new IllegalArgumentException("");
```

Comment:
we should have a better description here, even if it won't happen under any normal circumstance

Reply:
:))))